Issue-1146  part a (#14788)
terraform-providers/terraform-provider-aws: website/docs/r/emr_cluster.html.markdown

    instance_type_configs {
      bid_price_as_percentage_of_on_demand_price = 80
      ebs_config {
        size                 = 100
        type                 = "gp2"
        volumes_per_instance = 1
      }
      instance_type     = "m3.xlarge"
      weighted_capacity = 1
    }
    instance_type_configs {
      bid_price_as_percentage_of_on_demand_price = 100
      ebs_config {
        size                 = 100
        type                 = "gp2"
        volumes_per_instance = 1
      }
      instance_type     = "m4.xlarge"
      weighted_capacity = 1
    }
    instance_type_configs {
      bid_price_as_percentage_of_on_demand_price = 100
      ebs_config {
        size                 = 100
        type                 = "gp2"
        volumes_per_instance = 1
      }
      instance_type     = "m4.2xlarge"
      weighted_capacity = 2
    }
    launch_specifications {
      spot_specification {
        block_duration_minutes   = 0
        timeout_action           = "SWITCH_TO_ON_DEMAND"
        timeout_duration_minutes = 10
      }
    }
    name                      = "core fleet"
    target_on_demand_capacity = 2
    target_spot_capacity      = 2
  }
}

> resource "aws_emr_instance_fleet" "task" {

Reviewer: Oh ha, not sure how I missed this, but this resource doesn't exist, yet 😉 , so we should remove it from the example in case its not included in the same release.

  cluster_id          = aws_emr_cluster.example.id
  instance_fleet_type = "TASK"
  instance_type_configs {
    bid_price_as_percentage_of_on_demand_price = 100
    ebs_config {
      size                 = 100
      type                 = "gp2"
      volumes_per_instance = 1
    }
    instance_type     = "m4.xlarge"
    weighted_capacity = 1
  }
  instance_type_configs {
    bid_price_as_percentage_of_on_demand_price = 100
    ebs_config {
      size                 = 100
      type                 = "gp2"
      volumes_per_instance = 1
    }
    instance_type     = "m4.2xlarge"
    weighted_capacity = 2
  }
  launch_specifications {
    spot_specification {

>       block_duration_minutes   = 0

Reviewer: ```suggestion
      allocation_strategy      = "capacity-optimized"
      block_duration_minutes   = 0
```

guide for more information on these IAM roles. There is also a fully-bootable
example Terraform configuration at the bottom of this page.

## Instance Fleet

```hcl
resource "aws_emr_cluster" "example" {
  # ... other configuration ...
  master_instance_fleet {

>     instance_fleet_type = "MASTER"

Reviewer 2: when I tried to use this config, it failed with
```
Error: Unsupported argument

  on emr.tf line 23, in resource "aws_emr_cluster" "cluster":
  23:     instance_fleet_type = "MASTER"
```
after I removed "instance_fleet_type" it worked. Most likely you need to update this doc.
same for core and task

Author: yes, I just removed the `instance_fleet_type` for the instance_fleet and forget to update the document.

        size                 = 100
        type                 = "gp2"
        volumes_per_instance = 1
      }
      instance_type     = "m4.2xlarge"
      weighted_capacity = 2
    }
    launch_specifications {
      spot_specification {
        block_duration_minutes   = 0
        timeout_action           = "SWITCH_TO_ON_DEMAND"
        timeout_duration_minutes = 10
      }
    }
    name                      = "core fleet"
    target_on_demand_capacity = 2
    target_spot_capacity      = 2
  }
}
resource "aws_emr_instance_fleet" "task" {
  cluster_id          = aws_emr_cluster.example.id
  instance_fleet_type = "TASK"
  instance_type_configs {
    bid_price_as_percentage_of_on_demand_price = 100
    ebs_config {
      size                 = 100
      type                 = "gp2"
      volumes_per_instance = 1
    }
    instance_type     = "m4.xlarge"
    weighted_capacity = 1
  }
  instance_type_configs {
    bid_price_as_percentage_of_on_demand_price = 100
    ebs_config {
      size                 = 100
      type                 = "gp2"
      volumes_per_instance = 1
    }
    instance_type     = "m4.2xlarge"
    weighted_capacity = 2
  }
  launch_specifications {

>     spot_specification {

Reviewer: `allocation_strategy` is also missing from this example.

Author: yes, I just realized I added this based on the latest AWS API.
will fix it soon

Reviewer: (just noting here that its still missing with this resolved comment)

* `main_class` - (Optional) Name of the main class in the specified Java file. If not specified, the JAR file should specify a Main-Class in its manifest file.
* `properties` - (Optional) Key-Value map of Java properties that are set when the step runs. You can use these properties to pass key value pairs to your main function.

## master_instance_fleet Configuration Block

Supported nested arguments for the `master_instance_fleet` configuration block:

> * `instance_fleet_type` - (Required) has to be `MASTER`.

Reviewer: ```suggestion
```

* `instance_type_configs` - (Optional) Configuration block for instance fleet
* `launch_specifications` - (Optional) Configuration block for launch specification
* `target_on_demand_capacity` - (Optional)  The target capacity of On-Demand units for the instance fleet, which determines how many On-Demand instances to provision.
* `target_spot_capacity` - (Optional) The target capacity of Spot units for the instance fleet, which determines how many Spot instances to provision.
* `name` - (Optional) Friendly name given to the instance fleet.

## core_instance_fleet Configuration Block

Supported nested arguments for the `core_instance_fleet` configuration block:

> * `instance_fleet_type` - (Required) has to be `CORE`.

Reviewer: ```suggestion
```

* `launch_specifications` - (Optional) Configuration block for launch specification
* `target_on_demand_capacity` - (Optional)  The target capacity of On-Demand units for the instance fleet, which determines how many On-Demand instances to provision.
* `target_spot_capacity` - (Optional) The target capacity of Spot units for the instance fleet, which determines how many Spot instances to provision.
* `name` - (Optional) Friendly name given to the instance fleet.

## instance_type_configs Configuration Block

* `bid_price` - (Optional) The bid price for each EC2 Spot instance type as defined by `instance_type`. Expressed in USD. If neither `bid_price` nor `bid_price_as_percentage_of_on_demand_price` is provided, `bid_price_as_percentage_of_on_demand_price` defaults to 100%.
* `bid_price_as_percentage_of_on_demand_price` - (Optional) The bid price, as a percentage of On-Demand price, for each EC2 Spot instance as defined by `instance_type`. Expressed as a number (for example, 20 specifies 20%). If neither `bid_price` nor `bid_price_as_percentage_of_on_demand_price` is provided, `bid_price_as_percentage_of_on_demand_price` defaults to 100%.

> * `configurations` - (Optional) A configuration classification that applies when provisioning cluster instances, which can include configurations for applications and software that run on the cluster. List of `configuration` blocks.

Reviewer: These new configuration blocks also need documentation. 👍

* `properties` - (Optional) Key-Value map of Java properties that are set when the step runs. You can use these properties to pass key value pairs to your main function.

## master_instance_fleet Configuration Block

Supported nested arguments for the `master_instance_fleet` configuration block:

* `instance_type_configs` - (Optional) Configuration block for instance fleet
* `launch_specifications` - (Optional) Configuration block for launch specification
* `target_on_demand_capacity` - (Optional)  The target capacity of On-Demand units for the instance fleet, which determines how many On-Demand instances to provision.
* `target_spot_capacity` - (Optional) The target capacity of Spot units for the instance fleet, which determines how many Spot instances to provision.
* `name` - (Optional) Friendly name given to the instance fleet.

## core_instance_fleet Configuration Block

Supported nested arguments for the `core_instance_fleet` configuration block:

* `instance_type_configs` - (Optional) Configuration block for instance fleet

> * `launch_specifications` - (Optional) Configuration block for launch specification

Reviewer: Documentation for the new `launch_specifications` configuration block is missing, e.g. `spot_specification` configuration block and its arguments.

* `bid_price_as_percentage_of_on_demand_price` - (Optional) The bid price, as a percentage of On-Demand price, for each EC2 Spot instance as defined by `instance_type`. Expressed as a number (for example, 20 specifies 20%). If neither `bid_price` nor `bid_price_as_percentage_of_on_demand_price` is provided, `bid_price_as_percentage_of_on_demand_price` defaults to 100%.
* `configurations` - (Optional) A configuration classification that applies when provisioning cluster instances, which can include configurations for applications and software that run on the cluster. List of `configuration` blocks.
* `ebs_config` - (Optional) Configuration block(s) for EBS volumes attached to each instance in the instance group. Detailed below.
* `instance_type` - (Required) An EC2 instance type, such as m4.xlarge.
* `weighted_capacity` - (Optional) The number of units that a provisioned instance of this type provides toward fulfilling the target capacities defined in `aws_emr_instance_fleet`.

> 
> 
> ## configurations Configuration Block
> A configuration classification that applies when provisioning cluster instances, which can include configurations for applications and software that run on the cluster. See [Configuring Applications](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-configure-apps.html).

Reviewer: Our `markdown-lint` handlers were having some issues so CI didn't automatically report this, but spacing formatting here is a little off:

```suggestion

## configurations Configuration Block

A configuration classification that applies when provisioning cluster instances, which can include configurations for applications and software that run on the cluster. See [Configuring Applications](https://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-configure-apps.html).
```

> * `classification` - The classification within a configuration.
> 
> * `properties` - A set of properties specified within a configuration classification

Reviewer: Formatting and fix for set versus map terminology as the distinction is important in Terraform

```suggestion
* `classification` - (Optional) The classification within a configuration.
* `properties` - (Optional) A map of properties specified within a configuration classification
```
